main.go

```diff
@@ -81,21 +83,43 @@ func main() {
 	if zone == "" {
 		zone, err = getZone()
 		if err != nil {
-			fmt.Fprintf(os.Stderr, "Warning: failed to determine zone: %s\n", err)
+			err = fmt.Errorf("failed to determine zone: %s\n", err)
 			zone = ""
+			if *generateMeshId {
+				// Zone is required while using with generateMeshID.
+				fmt.Fprintf(os.Stderr, "Error: failed to determine zone: %s\n", err)
+				os.Exit(1)
+			}
+			fmt.Fprintf(os.Stderr, "Warning: failed to determine zone: %s\n", err)
 		}
 	}
 
+	// Generate clusterName from metadata server or from command-line
+	// arguments, with the latter taking preference.
+	cluster := *gkeClusterName
+	var clusterErr error
+	if cluster == "" {
+		cluster, clusterErr = getClusterName()
+		if clusterErr != nil {
+			if *generateMeshId {
+				// The metadata server would return an error when running on GCE VMs.
+				// gkeClusterName should be passed in when using generateMeshID in GCE
+				// VM cases.
+				fmt.Fprintf(os.Stderr, "Error: generate-mesh-id-experimental flag was supplied, but was unable to determine the current cluster from the metadata server with error: %s\n", err)
+				os.Exit(1)
+			}
+			clusterErr = fmt.Errorf("Warning: failed to determine GKE cluster name: %s\n", err)
```

Comment:
The text here was mostly already added in `getClusterName()`, which seems appropriate.

```suggestion
			clusterErr = fmt.Errorf("Warning: %s\n", err)
```

Reply:
Done.